pkg/gardenlet/controller/bastion/bastion_reconciler.go

```diff
@@ -0,0 +1,266 @@
+
+	bastion := &operationsv1alpha1.Bastion{}
+	if err := gardenClient.Client().Get(ctx, request.NamespacedName, bastion); err != nil {
+		if apierrors.IsNotFound(err) {
+			logger.Debug("Skipping because Bastion does not exist anymore")
+			return reconcile.Result{}, nil
+		}
+		logger.Infof("Unable to retrieve object from store: %v", err)
+		return reconcile.Result{}, err
+	}
+
+	if !controllerutils.BastionIsManagedByThisGardenlet(ctx, gardenClient.Client(), bastion, r.config) {
+		logger.WithField("bastion-seed", *bastion.Spec.SeedName).Debug("Skipping because Bastion is not managed by this gardenlet")
+		return reconcile.Result{}, nil
+	}
+
+	// retrieve Kubernetes client for Seed cluster
+	seedClient, err := r.clientMap.GetClient(ctx, keys.ForSeedWithName(*bastion.Spec.SeedName))
+	if err != nil {
+		return reconcile.Result{}, fmt.Errorf("failed to get seed client: %w", err)
+	}
+
+	// get Shoot for the bastion
+	shoot := gardencorev1beta1.Shoot{}
+	shootKey := kutil.Key(bastion.Namespace, bastion.Spec.ShootRef.Name)
+	if err := gardenClient.Client().Get(ctx, shootKey, &shoot); err != nil {
+		return reconcile.Result{}, fmt.Errorf("could not get shoot %v: %v", shootKey, err)
+	}
+
+	if bastion.DeletionTimestamp != nil {
+		err = r.cleanupBastion(ctx, logger, gardenClient.Client(), seedClient.DirectClient(), bastion, &shoot)
+	} else {
+		err = r.reconcileBastion(ctx, logger, gardenClient.Client(), seedClient.DirectClient(), bastion, &shoot)
+	}
+
+	if cause := extensionscontroller.ReconcileErrCause(err); cause != nil {
+		logger.Errorf("Reconciling failed: %v", cause)
+	}
+
+	return extensionscontroller.ReconcileErr(err)
+}
+
+func (r *reconciler) reconcileBastion(
+	ctx context.Context,
+	logger logrus.FieldLogger,
+	gardenClient client.Client,
+	seedClient client.Client,
+	bastion *operationsv1alpha1.Bastion,
+	shoot *gardencorev1beta1.Shoot,
+) error {
+	logger.Debug("Reconciling")
+
+	// ensure finalizer is set
+	if !controllerutil.ContainsFinalizer(bastion, finalizerName) {
+		if err := controllerutils.PatchAddFinalizers(ctx, gardenClient, bastion, finalizerName); err != nil {
+			return fmt.Errorf("failed ensure %q finalizer on bastion: %v", finalizerName, err)
+		}
+		// the patch above already triggers a reconcile, so we can stop here
+		return nil
+	}
+
+	// prepare extension resource
+	extBastion := newBastionExtension(bastion, shoot)
+	extensionsIngress := make([]extensionsv1alpha1.BastionIngressPolicy, len(bastion.Spec.Ingress))
+	for i, ingress := range bastion.Spec.Ingress {
+		extensionsIngress[i] = extensionsv1alpha1.BastionIngressPolicy{
+			IPBlock: ingress.IPBlock,
+		}
+	}
+
+	// create or update the bastion in the seed cluster
+	if _, err := controllerutil.CreateOrUpdate(ctx, seedClient, extBastion, func() error {
+		extBastion.Spec.UserData = createUserData(bastion)
+		extBastion.Spec.Ingress = extensionsIngress
+		extBastion.Spec.Type = *bastion.Spec.ProviderType
+
+		return nil
+	}); err != nil {
```

Comment:
I think we still need to set the operation and timestamp annotations here, similar to https://github.com/gardener/gardener/blob/7c4388ee8bb0eaecfbae451772dc2577df45cc3f/pkg/operation/botanist/component/extensions/network/network.go#L188-L189